From stable-bounces@linux.kernel.org  Thu Jul 14 16:42:44 2005
Date: Thu, 14 Jul 2005 16:40:04 -0700
From: Andrew Vasquez <andrew.vasquez@qlogic.com>
To: stable@kernel.org
Cc: Andrew Vasquez <andrew.vasquez@qlogic.com>, Michael Reed <mdr@sgi.com>
Subject: [PATCH] qla2xxx: Correct handling of fc_remote_port_add() failure case.

Correct handling of fc_remote_port_add() failure case.

Immediately return if fc_remote_port_add() fails to allocate
resources for the rport.  Original code would result in NULL
pointer dereference upon failure.

Reported-by: Michael Reed <mdr@sgi.com>

Signed-off-by: Andrew Vasquez <andrew.vasquez@qlogic.com>
Signed-off-by: Chris Wright <chrisw@osdl.org>
---

--- linux-2.6.12.2/drivers/scsi/qla2xxx/qla_init.c.orig	2005-07-14 15:14:14.000000000 -0700
+++ linux-2.6.12.2/drivers/scsi/qla2xxx/qla_init.c	2005-07-14 15:14:54.000000000 -0700
@@ -1914,9 +1914,11 @@ qla2x00_reg_remote_port(scsi_qla_host_t 
 		rport_ids.roles |= FC_RPORT_ROLE_FCP_TARGET;
 
 	fcport->rport = rport = fc_remote_port_add(ha->host, 0, &rport_ids);
-	if (!rport)
+	if (!rport) {
 		qla_printk(KERN_WARNING, ha,
 		    "Unable to allocate fc remote port!\n");
+		return;
+	}
 
 	if (rport->scsi_target_id != -1 && rport->scsi_target_id < MAX_TARGETS)
 		fcport->os_target_id = rport->scsi_target_id;
